model/CompatibilityCheckerService.php

```diff
@@ -63,43 +68,64 @@ function __construct()
+            }
+
+            if ($check === self::CHECK_OS && $validOs && $rule->compatible) {
+                return 1;
+            }
+
 
-        $osVersion = explode('.', $this->osVersion);
-        foreach ($this->compatibility as $rule) {
-            //os name
-            if ($rule->os === $this->os) {
-                //os Version
-                $validOs = true;
-                if ($rule->osVersion !== "") {
-                    foreach (explode('.', $rule->osVersion) as $key => $version) {
-                        if (!isset($osVersion[$key]) || $osVersion[$key] !== $version) {
-                            $validOs = false;
-                        }
-                    }
-                }
-
-                //browser validation
-                if (empty($rule->versions)) {
-                    // all versions work
-                    $isValid = true;
-                } else {
-                    // it is valid if the version is in the array
-                    // OR if the browser is chrome or firefox and it is a newer version than those in the array
-                    $isValid = in_array($browserVersion[0], $rule->versions)
-                        || (in_array($rule->browser,
-                                array('Chrome', 'Firefox')) && $browserVersion[0] > max($rule->versions));
-                }
-
-                if ($validOs && ($rule->browser === ""
-                        || $rule->browser === $this->browser && $isValid)
-                ) {
-                    return $rule->compatible;
-                }
+            //browser validation
+            if ($rule->browser == $this->browser && empty($rule->versions)) {
+                // all versions work
+                $validBrowser = true;
+            } elseif($rule->browser == $this->browser) {
+                // it is valid if the version is in the array
+                // OR if the browser is chrome or firefox and it is a newer version than those in the array
+                $validBrowser = in_array($browserVersion[0], $rule->versions)
+                    || (in_array($rule->browser, ['Chrome', 'Firefox']) && $browserVersion[0] > max($rule->versions));
+            } else {
+                $validBrowser = false;
+            }
+
+            if ($validOs && $validBrowser) {
+                return $rule->compatible;
             }
         }
+
         return 2;
     }
+
+    /**
+     * Standard version_compare threats that  5.2 < 5.2.0, 5.2 < 5.2.1, ...
+     *
+     * @param $ver1
+     * @param $ver2
+     * @param null|string @see http://php.net/manual/en/function.version-compare.php
+     * @return mixed
+     */
+    public function versionCompare($ver1, $ver2, $operator = null)
+    {
+        $ver1 = preg_replace('#(\.0+)+($|-)#', '', $ver1);
+        $ver2 = preg_replace('#(\.0+)+($|-)#', '', $ver2);
+        if ($operator === null) {
+            $result = version_compare($ver1, $ver2);
```

Comment:
In this specific comparison 10.12.5 and 10.12 should be equal. We won't test every fixes version of Osx for example